### PR TITLE
45 full collections

### LINKED ISSUE #45: Full Collections
- One array of collections
- Includes bullet instances

### PR DESCRIPTION
**Pull request by Octowl**
closes #45 
- collections now come back from the db with all their bullets attached.
- This means there is NO independent bullet array returned. Bullets are now accessible through `collection.bullets`

### CODE REVIEW

**Comment by Octowl** [comment]
@omriBernstein 
We merged this but would still like to review the implementation. Then we can merge your review notes as a separate PR.

File: app/models/collection.js [hunk 1]
```diff
@@ -28,12 +26,27 @@ class Collection {
             _.extend(this, props);
         }
     }
+
+    deserializeBullets(bulletInstances) {
+        this.bullets = this.bullets.map(bulletId => {
```
Comment (omriAckley) [inline app/models/collection.js:31]: Not crazy about `this.bullets` being "hybrid". It seems like `this.bullets` should either be an array of ids or an array of bullet objects. Or possibly for easier id lookup, and object where the keys are bullet ids and the values are bullet objects.

File: app/models/collection.js [hunk 2]
```diff
@@ -28,12 +26,27 @@ class Collection {
...
+
     addBullet(bullet, index) {
-        var index = index || this.bullets.length;   //so we can preserver ordering in collections.bullet array
-        this.bullets = this.bullets.slice(0, index).concat(bullet.id).concat(this.bullets.slice(index));
+        index = index || this.bullets.length; //so we can preserver ordering in collections.bullet array
+        this.bullets = this.bullets.slice(0, index).concat(bullet).concat(this.bullets.slice(index));
         if (bullet.collections.indexOf(this.id) < 0) bullet.collections.push(this.id)
         return Promise.all([this.save(), bullet.save()])
-        .catch(err => console.error('error ', err))
+            .catch(err => console.error('error ', err))
```
Comment (omriAckley) [inline app/models/collection.js:49]: This seems OK for now, but you might consider using something like [growl](https://github.com/marcorinck/angular-growl) for errors that are relevant to the end user.

File: app/models/collection.js [hunk 3]
```diff
@@ -60,7 +76,7 @@ class Collection {
             .then(convertToInstances)
             .catch(err => {
                 console.log('created new collection');
```
Comment (omriAckley) [inline app/models/collection.js:78]: In general, remove `console.log`s from master.

File: app/models/collection.js [hunk 4]
```diff
@@ -75,8 +91,8 @@ class Collection {
     }
 
 
-    static fetchAllWithBullets(props) {
-         return this.fetchAll(props) // andrew's refactoring comment
+    static fetchAllWithBullets(props) {     //can delete
```
Comment (omriAckley) [inline app/models/collection.js:94]: Does this comment mean this is dead code?

File: app/scripts/collections/collection.directive.js [hunk 5]
```diff
@@ -10,14 +10,14 @@ bulletApp.directive('collection', function($log){
...
                 scope.formattedTitle = formatTitle(scope.collection);
                 scope.muted = false;
                 scope.$evalAsync();
             })
             .catch($log.err);
 
-            scope.newBullet = new Bullet.Task()
+            scope.newBullet = new Bullet.Task();
 
             function formatTitle(collection) {
```
Comment (omriAckley) [inline app/scripts/collections/collection.directive.js:22]: Consider moving this formatting function to a factory, mainly to keep the directive code nice and lightweight. Ooh, you might even make a [filter](https://scotch.io/tutorials/building-custom-angularjs-filters).

File: app/scripts/collections/collection.directive.js [hunk 6]
```diff
@@ -45,15 +45,15 @@ bulletApp.directive('collection', function($log){
             scope.removeBullet = function(bullet) {
                 scope.collection.removeBullet(bullet)
                 .then(function(){
-                    scope.bullets = scope.bullets.filter(b => b.id !== bullet.id);
+                    scope.collection.bullets = scope.collection.bullets.filter(b => b.id !== bullet.id);
```
Comment (omriAckley) [inline app/scripts/collections/collection.directive.js:48]: Could this happen in the logic of `removeBullet`? If so, I'm thinking you might as well.

File: app/scripts/collections/collection.directive.js [hunk 7]
```diff
@@ -45,15 +45,15 @@ bulletApp.directive('collection', function($log){
...
                 .catch($log.err);
             };
+
             scope.addBullet = function(bullet) {
                 if (bullet.content.length > 0) {
                   scope.collection.addBullet(bullet)
                   .then(function(){
-                      scope.bullets.push(bullet);
                       scope.newBullet = new Bullet.Task()
                       scope.$evalAsync()
```
Comment (omriAckley) [inline app/scripts/collections/collection.directive.js:58]: Consider moving this digesting logic into the factory methods if possible, I'm thinking as a `$rootScope.$evalAsync`. This is just so you don't have to keep track of if/when to digest all over the place.

File: app/scripts/bullets/bullet.directive.js [hunk 8]
```diff
@@ -10,6 +10,7 @@ bulletApp.directive('bullet', function () {
             header: '@'
         },
         link: function (scope, element) {
```
Comment (omriAckley) [inline app/scripts/bullets/bullet.directive.js:12]: I'd say there's a good bit of logic going on in this directive that isn't directly related to managing the scope.

File: app/scripts/bullets/bullet.directive.js [hunk 9]
```diff
@@ -10,6 +10,7 @@ bulletApp.directive('bullet', function () {
...
+
             scope.typeDict = {
```
Comment (omriAckley) [inline app/scripts/bullets/bullet.directive.js:14]: Consider making this into a custom filter.

File: app/models/collection.js [hunk 10]
```diff
@@ -45,11 +58,14 @@ class Collection {
     }
 
     save() {
-        this.bullets = this.bullets.map(bullet => bullet.id);  //beforeSave, converts bullet instances to ids
-        return db.rel.save('collection', this);
+        const bulletInstances = this.bullets;
+        this.serializeBullets();
```
Comment (omriAckley) [inline app/models/collection.js:62]: Again, I find this kind of messy. If you had two properties on the collection—one for the bullet ids and one for the bullet instances, could that work?

File: app/models/collection.js [hunk 11]
```diff
@@ -28,12 +26,27 @@ class Collection {
             _.extend(this, props);
```
Comment (omriAckley) [inline app/models/collection.js:26]: If you're going with ES6 anyways, you might go with [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign).

**Pull request merged**